examples/hal/quad/main.rs

```diff
@@ -4,613 +4,1357 @@
 )]
 
 extern crate env_logger;
-extern crate gfx_hal as hal;
 #[cfg(feature = "dx12")]
 extern crate gfx_backend_dx12 as back;
-#[cfg(feature = "vulkan")]
-extern crate gfx_backend_vulkan as back;
-#[cfg(feature = "metal")]
-extern crate gfx_backend_metal as back;
 #[cfg(feature = "gl")]
 extern crate gfx_backend_gl as back;
-
-extern crate winit;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
 extern crate image;
+extern crate takeable_option;
+extern crate winit;
 
-use hal::{buffer, command, format as f, image as i, memory as m, pass, pso, pool};
-use hal::{Device, Instance, PhysicalDevice, Surface, Swapchain};
-use hal::{
-    DescriptorPool, FrameSync, Primitive,
-    Backbuffer, SwapchainConfig,
-};
-use hal::format::{AsFormat, ChannelType, Rgba8Srgb as ColorFormat, Swizzle};
-use hal::pass::Subpass;
-use hal::pso::{PipelineStage, ShaderStageFlags, Specialization};
-use hal::queue::Submission;
-
-use std::io::Cursor;
-
-const ENTRY_NAME: &str = "main";
-
-#[derive(Debug, Clone, Copy)]
-#[allow(non_snake_case)]
-struct Vertex {
-    a_Pos: [f32; 2],
-    a_Uv: [f32; 2],
-}
+#[macro_use]
+extern crate glsl_to_spirv_macros;
+#[macro_use]
+extern crate glsl_to_spirv_macros_impl;
 
-const QUAD: [Vertex; 6] = [
-    Vertex { a_Pos: [ -0.5, 0.33 ], a_Uv: [0.0, 1.0] },
-    Vertex { a_Pos: [  0.5, 0.33 ], a_Uv: [1.0, 1.0] },
-    Vertex { a_Pos: [  0.5,-0.33 ], a_Uv: [1.0, 0.0] },
+#[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal", feature = "gl"))]
+mod renderer {
```

Comment:
I agree with @msiglreith here that the increased complexity of the quad example is not a good thing. The examples need to be straightforward and readable, also the less code the better. Could you make this simpler?

Reply:
I wrapped many things in `#[cfg]`s to stop unused variable warnings, I can get rid of them if your ok with warnings on some/all builds.

Reply:
Do you mean the warnings shown when no backends are selected? We should automatically use the `empty` backend in this case.

Reply:
Yes, but then we'd have to put certain functions behind an feature gate, because they require certain functions not present in the empty backend. I've done so in the next commit and I can revert it if you feel it's more complex.